oops i idd it again: API change, pass down full components or use provided Question component

xaiki opened the pull request:
/!\ This is still missing documentation /!\

i wanted to have a check with you of what you think of this direction:
instead of passing just the label and constructing all the component in a
loop, we create a Question component that you can use instead of the <span/>
if you only want the previous API, or you can implement your own that just
needs to implement an onChange callback whenever the data is ready to be
added to values.

this allows to implement maps, typeahead, or any other kind of custom
components. we should probably ship some of those so that we have pretty and
useable stuff right away.

cheers,

xaiki committed:
indent all the things (and transform to PureComponents because it's free)

xaiki committed:
/!\ API change: introduce Question component and port questions API /!\

/!\ this breaks the children API /!\

instead of passing just the label and constructing all the component in a
loop, we create a Question component that you can use instead of the <span/>
if you only want the previous API, or you can implement your own that just
needs to implement an onChange callback whenever the data is ready to be
added to values.

this allows to implement maps, typeahead, or any other kind of custom
components. we should probably ship some of those so that we have pretty and
useable stuff right away.

theiliad commented:
I think this is a good change, not sure if you saw my comments here: https://github.com/GopherLabsLtd/react-eureka-form/pull/2

I'll merge soon

xaiki commented:
I think it should not be merged without doc, and as it's yet another API breakage, maybe we want to merge it with a few more question types too, and a bit more brain cycles.

xaiki commented:
(I had not seen your comments, answered there)

theiliad commented:
@xaiki Do you have some sample code of how you would use this new Question API?

Thanks.

xaiki committed:
doc for Component API

xaiki commented:
just pushed to the README.md, note that it's pseudo code, i'll write what i need for my current project tomorrow.

xaiki commented:
for future reference:

### Full API
In order to implement your own Question Helper you just need to call
`onChange` when your value changes, if you are going to use an `<input/>`
tag, please pass down the `type` property to make sure HTML5 validation
still works. You should also make sure to gracefully handle the `children`
prop, as it'll be usual to pass down the question component there.

```jsx
class InputQuestion extends React.PureComponent {
    render () {
        const { onChange, type, children } = this.props
        return (
            <div>
                <span>{children}</span>
                <input onChange={onChange} type={type}/>
            </div>
        )
    }
}


class ListQuestion extends React.PureComponent {
    render () {
        const { options, onChange, children } = this.props
        return (
            <ul>
                <h1>{children}</h1>
                { options.map(opt, i => (
                    <li key={i} onClick={() => onChange(opt)}>{opt}</li>
                )}
            </ul>
        )
    }
}

const MyForm = ({values = {}}) => (
    <EurekaForm onSubmit={doStuff}>
        <InputQuestion type='name'>
          What's your name ?
        </Question>
        <ListQuestion type='country' options={['Argentina', 'Brazil', 'Canada']}>
          Hello <b>{values.name}</b>, and your country ?
        </Question>
    </EurekaForm>
)
```

theiliad commented:
Cool, I like that idea actually!

xaiki committed:
README: component example is simpler with functional components

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
README: fixup: correctly close tags

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki commented:
this actually proved to be quite broken...
i'm in the process of re-writing it to work, need to split out all the code that selects the input, had to re-do the animation logic and am in the process of passing validation to the child too....

theiliad commented:
ok, animation logic could actually use some re-doing 😄 

Thanks!

xaiki committed:
stop accessing DOM directly do things the react way.

 - create a questionsRef object to hold a ref to all Questions DOM nodes
 - extract values from questionsRef
 - process count, total number and error from state and questionsRef
 - process current/next/previous classNames from state
 - drop all use of innerHTML
 - implement focus() method in Question
 - implement focus prop on question to set disabled attribute (may need to
   move to a method

xaiki committed:
Move validation to Question Component

we introduce a `validate()` method that defaults to using `_validateHTML5()`
(note:
         this breaks html5 detection, but there was no validation happening when
         HTML5 was not available anyway.
)

you are now *required* to provide a `this.inputRef` in the Question
Component for this to work, or re-implement your own `validate()` method
(you can still use the `_validateHTML5()` method if you wish by passing it
an `inputRef`).

xaiki committed:
Reimplement Animations (still needs fixing)

we completly rework how animation work, introducing a `.current`, `.next`
and `.prev` system of classNames, we also introduce a `Number` component
(that we export so you can re-use it in other parts of your UI) that uses
the same CSS as the questions roll to handle the counter.

we will animate any `<label/>` we find, so if you want something to be
roll-animated in your Component just wrap it in a `<label/>` (maybe we need
to className it ?), all other components will just disappear on transition.

xaiki committed:
indent style.css

xaiki committed:
remove debug messages

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki commented:
lots of patches =/ couldn't do it with less, and this still needs some love.
this patchset reworks a lot of the internal logic and the animation code.
there are still a few open issues that i'd like to discuss:
- we will animate *any* `<label>` we find in the `<Question/>` component maybe we need to className it ?
- we introduce a `validate()` and `focus()` method that require the `<Question/>` implementer to provide a `this.inputRef` to work. the 'correct' way to make that work would be to look for a `this` ref with ReactDOM, but that would introduce a new dependency, and probably make even harder to port our code to React Native.
- we have both a `focus()` method and a `focus` prop, that we use to focus on one side and disable the input on the other, it's a bit shady to export that to downstream users

theiliad commented:
@xaiki I was testing the PR today, couldn't get any of the sample code to work using your updated API. Is this a complete PR or a work in progress?

xaiki committed:
add simple example

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki commented:
it is *NOT* ready to merge, see previous comment for points that needs to be discussed. 
in the changes i did, i had to force `<Question>` to be a `React.Component` (instead of using functional components). the reason for that is the `validate()` and `focus()` issues stated before, they require a ref that require a full component.

that said in the last commit pushed here i've added an example that i'll keep up to date.
it currently has issues with CSS, as the `Question` `input` styling has to be independent and i couldn't find a nice way to do that (help wanted)

cheers.

xaiki commented:
I've been banging my head all day around this approach, and while i think it has a lot of good points, i'm not sure i can come up with an implementation, i'm now spliting out `<input>` in an `Input` component to style it, and need to pass down a lot of methods from Question that break generalizability, maybe it needs to go back to the drawing board...
